crates/proto/src/lib.rs

```diff
@@ -88,3 +88,61 @@ pub use crate::xfer::retry_dns_handle::RetryDnsHandle;
 #[cfg(feature = "backtrace")]
 pub use error::{ENABLE_BACKTRACE, ExtBacktrace};
 pub use error::{ForwardData, ForwardNSData, ProtoError, ProtoErrorKind};
+
+#[cfg(not(feature = "no-std-rand"))]
+pub(crate) use rand::random;
+
+#[cfg(feature = "no-std-rand")]
+pub(crate) use no_std_rand::random;
+#[cfg(feature = "no-std-rand")]
+pub use no_std_rand::seed;
+
+/// A simple shim that allows us to use a [`StdRng`] in `no_std` environments.
+#[cfg(feature = "no-std-rand")]
+mod no_std_rand {
+    use core::cell::RefCell;
+
+    use critical_section::Mutex;
+    use once_cell::sync::Lazy;
+    use rand::distr::{Distribution, StandardUniform};
+    use rand::{Rng, SeedableRng, rngs::StdRng};
+
+    /// Generates a random value on `no_std`.
+    ///
+    /// ** WARNING **
+    /// The random value is predictable, unless seeded using [`crate::seed`]!
+    ///
+    /// Depending on the usage of this library, this may yield predictable DNS requests that attackers can
+    /// use to feed wrong responses to hickory.
+    /// Always seed this library before using in `no_std` environments.
+    pub(crate) fn random<T>() -> T
+    where
+        StandardUniform: Distribution<T>,
+    {
+        critical_section::with(|cs| {
+            RNG.borrow_ref_mut(cs)
+                .as_mut()
+                .expect("The no_std rng was not seeded using `hickory_proto::seed()`")
+                .random()
+        })
+    }
+
+    /// Seed the rng that is used to create random DNS IDs throughout the lib (no_std-only).
+    pub fn seed(seed: u64) {
+        critical_section::with(|cs| *RNG.borrow_ref_mut(cs) = Some(StdRng::seed_from_u64(seed)));
+    }
+
+    static RNG: Lazy<Mutex<RefCell<Option<StdRng>>>> = Lazy::new(|| Mutex::new(RefCell::new(None)));
```

Comment:
I think we might not need the `Lazy` here? Both `Mutex::new()` and `RefCell::new()` are `const`.